FIAT/finite_element.py

```diff
@@ -27,6 +28,82 @@
 
 
 class FiniteElement(object):
+    """Class implementing a basic abstraction template for special
+    types of finite elements, like traces. Most special methods will
+    need to be implemented by the special element in its module."""
+
+    def __init__(self, ref_el, order, formdegree=None, mapping="affine"):
+
+        self.order = order
+        self.formdegree = formdegree
+        self.mapping = mapping
+
+        self.ref_el = ref_el
+
+    def degree(self):
+        "Return the degree of the (embedding) polynomial space."
+        raise NotImplementedError()
+
+    def get_order(self):
+        "Return the order of the element (may be different from the degree)"
+        return self.order
+
+    def entity_dofs(self):
+        """Return the map of topological entities to degrees of
+        freedom for the finite element."""
+        raise NotImplementedError()
+
+    def entity_closure_dofs(self):
+        """Return the map of topological entities to degrees of
+        freedom on the closure of those entities for the finite element."""
+        raise NotImplementedError()
+
+    def get_reference_element(self):
+        "Return the reference element for the finite element."
+        return self.ref_el
+
+    def get_nodal_basis(self):
+        """Return the nodal basis, encoded as a PolynomialSet object,
+        for the finite element."""
+        raise NotImplementedError()
+
+    def get_coeffs(self):
```

Comment:
I'm not sure it's used too much outside of FIAT. It is implemented in `CiarletElement` out of the constructed `poly_set`